src/theory/datatypes/inference_manager.cpp

```diff
@@ -0,0 +1,190 @@
+    TypeNode tn = n[0].getType();
+    if (!tn.isDatatype())
+    {
+      addLemma = true;
+    }
+    else
+    {
+      const DType& dt = tn.getDType();
+      addLemma = dt.involvesExternalType();
+    }
+  }
+  else if (n.getKind() == LEQ || n.getKind() == OR)
+  {
+    addLemma = true;
+  }
+  if (addLemma)
+  {
+    Trace("dt-lemma-debug") << "Communicate " << n << std::endl;
+    return true;
+  }
+  Trace("dt-lemma-debug") << "Do not need to communicate " << n << std::endl;
+  return false;
+}
+
+void InferenceManager::process()
+{
+  // process pending lemmas, used infrequently, only for definitional lemmas
+  doPendingLemmas();
+  // now process the pending facts
+  size_t i = 0;
+  NodeManager* nm = NodeManager::currentNM();
+  while (!d_theoryState.isInConflict() && i < d_pendingFact.size())
+  {
+    std::pair<Node, Node>& pfact = d_pendingFact[i];
+    Node fact = pfact.first;
+    Node exp = pfact.second;
+    Trace("datatypes-debug")
+        << "Assert fact (#" << (i + 1) << "/" << d_pendingFact.size() << ") "
+        << fact << " with explanation " << exp << std::endl;
+    // check to see if we have to communicate it to the rest of the system
+    if (mustCommunicateFact(fact, exp))
+    {
+      Node lem = fact;
+      if (exp.isNull() || exp == d_true)
+      {
+        Trace("dt-lemma-debug") << "Trivial explanation." << std::endl;
+      }
+      else
+      {
+        Trace("dt-lemma-debug") << "Get explanation..." << std::endl;
+        std::vector<TNode> assumptions;
+        explain(exp, assumptions);
+        if (assumptions.empty())
+        {
+          lem = fact;
+        }
+        else
+        {
+          std::vector<Node> children;
+          for (const TNode& assumption : assumptions)
+          {
+            children.push_back(assumption.negate());
+          }
+          children.push_back(fact);
+          lem = nm->mkNode(OR, children);
+        }
+      }
+      Trace("dt-lemma") << "Datatypes lemma : " << lem << std::endl;
+      doSendLemma(lem);
+    }
+    else
+    {
+      // assert the internal fact
+      bool polarity = fact.getKind() != NOT;
+      TNode atom = polarity ? fact : fact[0];
+      assertInternalFact(atom, polarity, exp);
+      d_addedFact = true;
+    }
+    Trace("datatypes-debug") << "Finished fact " << fact
+                             << ", now = " << d_theoryState.isInConflict()
+                             << " " << d_pendingFact.size() << std::endl;
+    i++;
+  }
+  d_pendingFact.clear();
+}
+
+bool InferenceManager::hasAddedFact() const
+{
+  return d_addedFact;
+}
+bool InferenceManager::hasAddedLemma() const
+{
+  return d_addedLemma;
+}
+bool InferenceManager::doSendLemma(Node lem, LemmaProperty p, bool cached)
+{
+  // don't cache lemmas with non-standard properties
+  Assert (!cached || p==LemmaProperty::NONE);
+  bool doSend = false;
+  if (!cached)
+  {
+    // always send
+    doSend = true;
+  }
+  else if (d_lemmasSent.find(lem) == d_lemmasSent.end())
```

Comment:
I am wondering if we should move the caching this to the buffered inference manager (together with a boolean that enables/disables caching of lemmas).

Reply:
I had considered this, however, that caching varies per theory.

The issue is that you must also be careful about LemmaProperty.  This leads to two cases:
(1) Theories don't use LemmaProperty and thus can cache uniformly `Node` only (for performance),
(2) Theories use LemmaProperty and must cache based on `std::pair<Node, LemmaProperty>`. 

Moreover, theories may know more efficient ways of caching e.g. non-linear arithmetic maintains two lemma caches for LemmaProperty={none, preprocess}. 

In short, I think caching isn't clear enough of a common feature since each theory has its own justifiably preferred way of doing it (in my estimation).  We could consider standardizing this later as the code develops.

Reply:
I support @4tXJ7f. Implementing this cache in a class derived from buffered inference manager is incredibly difficult and error-prone as it really needs to know what exactly its base class is doing. The cache should contain the lemmas that were sent to the output channel, and this is virtually impossible to get right if there is an abstraction layer in between that caches lemmas and may even drop lemmas.

As for the different caches for different lemma properties: I'm not to fond of this solution anyway, as it paves the way to adding the same lemma multiple times.

Reply:
Ok, given that it is critical to non-linear arithmetic at the moment, I'm on board with standardizing this now. Perhaps we should meet this week to discuss.

@nafur Regarding adding the same lemma with multiple properties several times, this is actually a question of model soundness, since it may be incorrect to drop a lemma with a different property (as a simple example say a lemma is sent with NONE and then afterwards with NEEDS_JUSTIFY). Its required in some cases to ensure this is done correctly.